### PR TITLE
ehancement(datadog_metrics sink, datadog_logs sink): improve retry behavior code quality

### PR DESCRIPTION
**Pull request by dsmith3197**
This doesn't actually change the retry behavior but cleans up the logic to make it clearer what status codes are retried, improves the error messaging, and deletes a dead code path.

https://datadoghq.atlassian.net/browse/OPW-245

### CODE REVIEW

File: src/sinks/datadog/metrics/service.rs [hunk 1]
```diff
@@ -31,23 +31,6 @@ impl RetryLogic for DatadogMetricsRetryLogic {
     fn is_retriable_error(&self, error: &Self::Error) -> bool {
         error.is_retriable()
     }
-
-    fn should_retry_response(&self, response: &Self::Response) -> RetryAction {
```
Comment (dsmith3197) [inline src/sinks/datadog/metrics/service.rs:35]: This is a dead code path because we call `DatadogApiError::from_result(result)` below.

Comment (neuronull) [reply src/sinks/datadog/metrics/service.rs:35]: 🤔  I'm missing how calling the `from_result` function makes this dead code path? The RetryLogic implementation is used in the service framework code ... Like, I see in `is_retriable_error()` , we are using the DatadogApiError. But in this function we aren't.

Comment (dsmith3197) [reply src/sinks/datadog/metrics/service.rs:35]: `DatadogApiError::from_result(result)` converts `result` into an error unless the status code is 2xx. This means that `should_retry_response` will never be call on a `response` that is non-2xx. Rather, the non-2xx status codes are converted into errors and the retry logic is handled in `error.is_retriable()`.

Comment (neuronull) [reply src/sinks/datadog/metrics/service.rs:35]: Ah, got it now. thanks for the added context. Would it be helpful maybe to add a comment here summarizing this? It seems like a not common thing we are doing for this sink.

**Review by neuronull — APPROVED** [review]
I'm suspecting we don't have great test coverage on the error handling code paths effected here but this seems reasonable to me.

**Pull request merged**